base/operators.jl

```diff
@@ -1078,8 +1078,7 @@ struct Fix1{F,T} <: Function
     f::F
     x::T
 
-    Fix1(f::F, x::T) where {F,T} = new{F,T}(f, x)
-    Fix1(f::Type{F}, x::T) where {F,T} = new{Type{F},T}(f, x)
+    Fix1(f, x) = new{Core.Typeof(f),Core.Typeof(x)}(f, x)
```

Comment:
This usage of `Core.Typeof` is not correct, because `Typeof` doesn't always return legal types:

```julia
julia> struct Fix1{F,T} <: Function
           f::F
           x::T

           Fix1(f, x) = new{Core.Typeof(f),Core.Typeof(x)}(f, x)
       end

julia> Fix1(==, Vector{Core._typevar(:T, Union{}, Any)})
ERROR: TypeError: in new, expected DataType, got Type{Fix1{typeof(==), Type{Array{T, 1}}}}
Stacktrace:
 [1] Fix1(f::Function, x::Type)
   @ Main ./REPL[3]:5
 [2] top-level scope
   @ REPL[4]:1
```

Reply:
Can we use it for `f`? (I see `ComposedFunction(outer, inner) = new{Core.Typeof(outer),Core.Typeof(inner)}(outer, inner)` above)

Reply:
I don't think it's possible to define methods on incomplete types with free typevars, but it will change the kind of error thrown to a `TypeError` during construction instead of a `MethodError` when actually called, so it's not great either

Reply:
I tried `f(::Type{T}) where {T} = T`, but `f(Vector{Core._typevar(:T, Union{}, Any)})` throws a `ERROR: UndefVarError: T not defined`
So we need something like:
```julia
_typeof(x) = typeof(x)
_typeof(::Type{T}) where {T} = @isdefined(T) ? Type{T} : DataType
```
?

Reply:
Is there anything we should talk at this point, or is this PR ready for merge?

Reply:
Merged, thanks